### PR TITLE
Stop using unexported function for obtaining symbol addresses

### PR DESCRIPTION
**Pull request by logankonopatzke**
In reference to issue #27, an alternative to kallsyms_lookup_name is now being used to locate symbols. get_kernel_symbol_address scans through the kernel .text section and obtains the symbol address. This method is significantly slower and uses much more CPU at first, but it shouldn't matter much since the symbols are resolved only in the module entry point. I found the kernel text section start and end (denoted by _text and _etext) via dumping /proc/kallsyms, but I'm not 100% that these addresses are the exact same on all systems.

### CODE REVIEW

**Comment by h33p** [comment]
Kernel performs ASLR on every boot (and this feature is enabled on almost all mainstream kernels), thus the addresses defined should not even match between boots. For instance, my kernel's text base is at `0xffffffff86000000` right now. I appreciate the effort, and if the kernel base gets retrieved dynamically, I will be able to merge the code in. Thank you.

**Comment by logankonopatzke** [comment]
🤦 Oh my, I forgot about KASLR, my bad! I'll find a way to do it dynamically. (and actually do more thorough testing this time lol)

**Comment by logankonopatzke** [comment]
Ok, I have developed a solution, but its not pretty. Resolving the symbols takes WAY longer (around 20 seconds in total for me), however, it seems to work. I tested on 2 different systems and the results come back as expected. The only thing I am concerned about is the order of the kernel sections being different on some systems.

File: kmem.c [hunk 1]
```diff
@@ -58,16 +59,19 @@ static void vmreadexit(void)
...
 static unsigned long get_kernel_symbol_address(const char *name)
 {
-	const unsigned long text_section_start = 0xffffffff98e00000;
-	const unsigned long text_section_end = 0xffffffff99c00e81;
+	const unsigned long in_data_section = (unsigned long)&jiffies; // Exported variable in the data section (D) used for counting ticks
 
+	bool found_symbol = false;
 	unsigned long offset;
-	for (offset = text_section_start; offset <= text_section_end; offset += 0x1)
+	for (offset = in_data_section; offset >= 0; offset -= 0x1)
```
Comment (h33p) [inline kmem.c:66]: Pointers would always be aligned to the native word length, so it is not needed to step by just one single value. Stepping by `sizeof(long)` would speed things up by 4-8 times

File: kmem.c [hunk 2]
```diff
@@ -58,16 +59,19 @@ static void vmreadexit(void)
...
 	{
 		char buffer[256];
 		snprintf(buffer, 256, "%pf", (void *)offset);
 		if (!strcmp(name, buffer))
-			return offset;
+			found_symbol = true; // We have found the function, now we just need to find the start of it
+		else if (found_symbol)
+			return offset += 0x1;
```
Comment (h33p) [inline kmem.c:73]: The increment is rather ambiguous and it is not common knowlege whether offset will be pre-incremented or post-incremented.

**Comment by h33p** [comment]
Thanks for the changes, I checked the code out, and it hangs my system. It is rather creative, not gonna lie, but does it have a place in a library that is used by hundreds of people? In kernel-level code? Don't get me wrong, I appreciate the effort, and I'm sure we can work something out. Now, I have left a few comments on the code, but something different is needed. Just recently, I have written a [kallsyms library](https://github.com/h33p/kallsyms-lp). Feel free to pull it in (preferably as a git submodule), or create a new solution that is stable. Good luck!

**Comment by logankonopatzke** [comment]
I really appreciate the feedback. It's probably best if I just rethink this and come back with something more practical. I'll definitely look into your livepatch library, it seems like a much more elegant solution.

**Pull request closed without merging**